## ひろしまQuest2020#stayhome【コース予測部門】

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 300)

### 管理番号

In [2]:
model_No = 20

### sub-modelを使用するかどうか

In [3]:
use_sub_model = True

In [4]:
ALL_MERGE = 'intermediate/all_merge_{}.f'.format(model_No)
SUBMIT = 'submit/course_{}.csv'.format(model_No)

if use_sub_model:
    ALL_MERGE = 'intermediate/all_merge_{}_sub.f'.format(model_No)
    SUBMIT = 'submit/course_{}_sub.csv'.format(model_No)
else:
    OUT_SUBMODEL = 'intermediate/course_predict_{}.f'.format(model_No)

In [5]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 187)

In [6]:
if use_sub_model:
    all_pitch['predict_curve'] = all_pitch['predict_curve'] / all_pitch['predict_straight']
    all_pitch['predict_slider'] = all_pitch['predict_slider'] / all_pitch['predict_straight']
    all_pitch['predict_shoot'] = all_pitch['predict_shoot'] / all_pitch['predict_straight']
    all_pitch['predict_fork'] = all_pitch['predict_fork'] / all_pitch['predict_straight']
    all_pitch['predict_changeup'] = all_pitch['predict_changeup'] / all_pitch['predict_straight']
    all_pitch['predict_sinker'] = all_pitch['predict_sinker'] / all_pitch['predict_straight']
    all_pitch['predict_cutball'] = all_pitch['predict_cutball'] / all_pitch['predict_straight']
    all_pitch.drop(columns=[
        'predict_0', 'predict_1', 'predict_2', 'predict_3', 'predict_4', 'predict_5', 'predict_6',
        'predict_7', 'predict_8', 'predict_9', 'predict_10', 'predict_11', 'predict_12',
        'predict_straight'
    ], inplace=True)

In [7]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 173)


No  ball  course  pitch_cnt_in_game  inning  bat_cnt_in_inning  \
0   0   0.0     0.0                  1       1                  1   
1   1   0.0     8.0                  2       1                  1   
2   2   0.0     5.0                  3       1                  1   
3   3   0.0    12.0                  4       1                  2   
4   4   0.0     8.0                  5       1                  2   

   pitch_cnt_in_bat  pitch_order  player_cnt_in_game  pitcher_cnt_in_game  \
0                 1            1                   1                    1   
1                 2            1                   1                    2   
2                 3            1                   1                    3   
3                 1            1                   2                    4   
4                 2            1                   2                    5   

   pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  home_point  away_point  \
0                      1          1                1           0           0   
1                      2          1                1           0           0   
2                      3          1                1           0           0   
3                      4          2                1           0           0   
4                      5          2                1           0           0   

   out_cnt  ball_cnt  strike_cnt  pitch_LR  bat_LR  bc_curve  bc_slider  \
0        0         0           0       0.0     1.0  0.237468   0.325101   
1        0         0           1       0.0     1.0  0.210923   0.247198   
2        0         0           2       0.0     1.0  0.138758   0.130040   
3        1         0           0       0.0     0.0  0.213866   0.474931   
4        1         0           1       0.0     0.0  0.215241   0.420099   

   bc_shoot   bc_fork  bc_changeup  bc_sinker  bc_cutball  bc_high_str  \
0  0.209547  0.145227     0.092978   0.027829    0.113620     0.135883   
1  0.177502  0.320359     0.147952   0.039943    0.116772     0.106283   
2  0.092263  0.419906     0.117327   0.038140    0.046858     0.076007   
3  0.139964  0.088519     0.038543   0.007023    0.138168     0.151308   
4  0.145723  0.196762     0.051566   0.009680    0.118444     0.113468   

   bc_high_ball  bc_mid_str  bc_low_str  bc_low_ball  bc_left_str  \
0      0.221489    0.204959    0.174254     0.263415     0.242045   
1      0.230235    0.157397    0.138195     0.367890     0.184621   
2      0.284249    0.102015    0.095055     0.442674     0.118315   
3      0.196937    0.206965    0.157643     0.287147     0.105057   
4      0.206297    0.162901    0.127988     0.389347     0.084346   

   bc_left_ball  bc_center_str  bc_right_str  bc_right_ball  first  second  \
0      0.294634       0.157681      0.115370       0.190270      0       0   
1      0.355720       0.111331      0.105923       0.242405      0       0   
2      0.488095       0.076190      0.078571       0.238828      0       0   
3      0.153718       0.164756      0.246104       0.330367      0       0   
4      0.162167       0.120564      0.199445       0.433477      0       0   

   third  base_cnt  top_bot  role  pos_pit  date_from_opening  elapsed_min  \
0      0         0        0     1        0                  0     0.000000   
1      0         0        0     1        0                  0     0.200000   
2      0         0        0     1        0                  0     0.383333   
3      0         0        0     1        0                  0     1.016667   
4      0         0        0     1        0                  0     1.200000   

   min_diff  total_inning  elapsed_from_inning  ballnum_from_inning  \
0       NaN             0             0.000000                    1   
1  0.200000             0             0.200000                    2   
2  0.183333             0             0.383333                    3   
3  0.633333             0             1.016667                    4   
4  0.183333             0             1.

### train
- 行数: 257117

In [8]:
train = all_pitch.dropna(subset=['course'])
train.shape

(257117, 173)

### test
- 行数: 521650

In [9]:
test = all_pitch[all_pitch['course'].isnull()]
test.shape

(521650, 173)

In [10]:
del all_pitch
gc.collect()

40

In [11]:
train_d = train.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

In [12]:
test_d = test.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

## Dataset作成

In [13]:
lgb_train = lgb.Dataset(train_d, train['course'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [14]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 13,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 8.769293390201968, 
        'lambda_l2': 3.913949617576324e-05, 
        'num_leaves': 6, 
        'feature_fraction': 0.4, 
        'bagging_fraction': 0.8391111798378441, 
        'bagging_freq': 4, 
        'min_child_samples': 50
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
best_cv_score = cv_results['multi_logloss-mean'][-1]
print('Best CV score:', best_cv_score)

[100]	cv_agg's multi_logloss: 2.32775 + 0.00100935
[200]	cv_agg's multi_logloss: 2.3208 + 0.000969677
[300]	cv_agg's multi_logloss: 2.31828 + 0.000939539
[400]	cv_agg's multi_logloss: 2.31676 + 0.000975521
[500]	cv_agg's multi_logloss: 2.31601 + 0.000856438
[600]	cv_agg's multi_logloss: 2.31571 + 0.000896158
[700]	cv_agg's multi_logloss: 2.31543 + 0.000907502
[800]	cv_agg's multi_logloss: 2.31538 + 0.000925921
Best num_boost_round: 793
Best CV score: 2.315349281506766
CPU times: user 1h 11min 45s, sys: 7min 35s, total: 1h 19min 20s
Wall time: 9min 57s


In [15]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [16]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 21min 56s, sys: 2min 15s, total: 24min 11s
Wall time: 3min 1s


### Feature Importance

In [17]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
119            left_ball       612
125           weight_bat       630
114            high_ball       632
6    pitcher_cnt_in_game       644
137      salary_year_bat       656
122           right_ball       671
45   elapsed_from_inning       694
115              mid_str       715
50    ave_elapsed_batter       749
113             high_str       784
139              bmi_bat       784
116              low_str       804
121            right_str       806
43              min_diff       828
118             left_str       831
140           batter_cnt       842
120           center_str       844
51   diff_elapsed_batter       879
53     diff_elapsed_game       899
52   diff_elapsed_inning       992
168       predict_sinker      1008
48      ave_elapsed_game      1034
163        predict_curve      1096
49    ave_elapsed_inning      1121
167     predict_changeup      1220
164       predict_slider      1223
41     date_from_opening      1248
169      predict_cutball      1288
165        predict_shoot      1506
166         predict_fork      1557

In [18]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 5min 39s, sys: 3.1 s, total: 5min 42s
Wall time: 45.7 s


## 結果出力

In [19]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 14)


index         0         1         2         3         4         5  \
0      0  0.079762  0.132805  0.074496  0.040194  0.053744  0.031691   
1      1  0.074360  0.137558  0.076428  0.039517  0.048858  0.028300   
2      2  0.031615  0.086602  0.057300  0.026069  0.037209  0.024249   
3      3  0.046554  0.140215  0.070892  0.027175  0.035031  0.025483   
4      4  0.062344  0.138268  0.090329  0.032394  0.037113  0.041110   

          6         7         8         9        10        11        12  
0  0.058046  0.050055  0.033072  0.184248  0.093212  0.080730  0.087945  
1  0.053359  0.056816  0.038825  0.156357  0.103790  0.086075  0.099758  
2  0.040898  0.054342  0.029299  0.166274  0.105186  0.191816  0.149141  
3  0.037712  0.050542  0.036936  0.162148  0.092913  0.130153  0.144246  
4  0.045631  0.067154  0.033093  0.132014  0.060336  0.119945  0.140268

In [20]:
submit.to_csv(SUBMIT, header=False, index=False)

### 球種予測で使用する

In [21]:
%%time
if not use_sub_model:
    train_predict = lgb_model.predict(train_d, num_iteration = lgb_model.best_iteration)

CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 7.87 µs


In [22]:
if not use_sub_model:
    df_train_predict = pd.DataFrame(train_predict).reset_index()
    submodel = pd.concat([df_train_predict, submit], ignore_index=True)
    submodel.drop(columns=['index'], inplace=True)
    submodel.rename(columns={
        0: 'predict_0',
        1: 'predict_1',
        2: 'predict_2',
        3: 'predict_3',
        4: 'predict_4',
        5: 'predict_5',
        6: 'predict_6',
        7: 'predict_7',
        8: 'predict_8',
        9: 'predict_9',
        10: 'predict_10',
        11: 'predict_11',
        12: 'predict_12'
    }, inplace=True)
    print(submodel.shape)
    submodel.head()

In [23]:
if not use_sub_model:
    submodel.to_feather(OUT_SUBMODEL)

### SIGNATE

In [24]:
'signate submit --competition-id=276 ./{} --note feat={}_cv={}'.format(SUBMIT, column_cnt, best_cv_score)

'signate submit --competition-id=276 ./submit/course_20_sub.csv --note feat=173_cv=2.315349281506766'

### GCS

In [25]:
'gsutil cp ./{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ./submit/course_20_sub.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 20_sub(173)
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 793
- Best CV score: 2.315349281506766
- Wall time: 9min 57s
- 評価結果  : 2.3585163

#### 20(166) ... player同士の組み合わせを削除
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 572
- Best CV score: 2.3271237073894153
- Wall time: 6min 59s
- 評価結果  : 2.3552697

#### 19_sub(197) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 774
- Best CV score: 2.316818574234304
- Wall time: 10min 49s
- 評価結果  : 2.3581029

#### 19(190) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 550
- Best CV score: 2.327580481542456
- Wall time: 7min 36s
- 評価結果  : 2.3554862

#### 18_sub(225) ...サブモデルで球種の予測を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 548
- Best CV score: 2.308116414786115
- Wall time: 8min 25s
- 評価結果  : 2.3638683

#### 18(217) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 373
- Best CV score: 2.3204478805445055
- Wall time: 6min 9s
- 評価結果  : 2.3604751

#### 17(192) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 415
- Best CV score: 2.3200530570975535
- Wall time: 15min 10s
- 評価結果  : 2.3607863

#### 16(181)  ...選手IDごとの2017年のコース実績
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 355
- Best CV score: 2.3203553166835866
- Wall time: 5min 8s
- 評価結果  : 2.3608998

#### 15(153)  ...左右で球種/コースを別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 932
- Best CV score: 2.3387434142884445
- Wall time: 9min 41s
- 評価結果  : 2.3552543 ★

#### 14(153)  ...左右で球種を別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 909
- Best CV score: 2.340082741044867
- Wall time: 9min 40s
- 評価結果  : 2.3555548

#### 3(157)  ...前の投球からの時間差を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 857
- Best CV score: 2.340703597440808
- Wall time: 9min 20s
- 評価結果  : 2.3554087

#### 2(156)  ...merge12のパラメータをチューニング
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 835
- Best CV score: 2.3419206656914
- Wall time: 9min 7s
- 評価結果  : 2.3555299

#### 2(156)  ...merge12(2017のコースデータを付加) 
- 'learning_rate' : 0.1,
- Best num_boost_round: 121
- Best CV score: 2.344104226715559
- Wall time: 12min 30s
- 評価結果  : 2.3564326

#### 1(143)  ...球種予測の11と同じデータ 
- 'learning_rate' : 0.1,
- Best num_boost_round: 122
- Best CV score: 2.3445089434874253
- Wall time: 12min 2s (local)
- 評価結果  : 2.3562648